pkg/controller/certificates/issuing/secret_manager.go

```diff
@@ -0,0 +1,222 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package issuing
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	"github.com/jetstack/cert-manager/pkg/controller/certificates/internal/secretsmanager"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+)
+
+// ensureSecretData ensures that the Certificate's Secret is up to date with
+// non-issuing condition related data. Currently only reconciles on Annotations
+// and Labels from the Certificate's SecretTemplate.
+func (c *controller) ensureSecretData(ctx context.Context, log logr.Logger, crt *cmapi.Certificate) error {
+	dbg := log.V(logf.DebugLevel)
+
+	// Retrieve the Secret which is associated with this Certificate.
+	secret, err := c.secretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
+
+	// Secret doesn't exist so we can't do anything. The Certificate will be
+	// marked for a re-issuance and the resulting Secret will be evaluated again.
+	if apierrors.IsNotFound(err) {
+		dbg.Info("secret not found", "error", err.Error())
+		return nil
+	}
+
+	// This error is transient, return error to be retried on the rate limiting
+	// queue.
+	if err != nil {
+		return err
+	}
+
+	secret = secret.DeepCopy()
+
+	var data secretsmanager.SecretData
+	if secret.Data != nil {
+		data = secretsmanager.SecretData{
+			PrivateKey:  secret.Data[corev1.TLSPrivateKeyKey],
+			Certificate: secret.Data[corev1.TLSCertKey],
+			CA:          secret.Data[cmmeta.TLSCAKey],
+		}
+	}
+
+	log = log.WithValues("secret", secret.Name)
+
+	// Check whether the Certificate's SecretTemplate matches that on the Secret.
+	secretTemplateMatchManagedFields, err := c.secretTemplateMatchesManagedFields(crt, secret, data)
+	if err != nil {
+		// An error here indicates that the managed fields are malformed and the
+		// decoder doesn't understand the managed fields on the Secret, or the
+		// signed certificate data could not be decoded. There is nothing more the
+		// controller can do here, so we exit nil so this controller doesn't end in
+		// an infinite loop.
+		log.Error(err, "failed to determine whether the SecretTemplate matches Secret")
```

Comment:
In the context of a human operator of cert-manager, what would this result in? Let me know wether or not my understanding is correct:

The issuing controller loop will always iterate over all Certificate-related secrets.
* What is the frequency of the loop?

If a secret is malformed and we reach this branch: there will be a log statement in the controller output (assuming log level includes `error`) and this log statement would repeat on each loop?

Reply:
The issuing controller will reconcile over Certificates when:
1. a Certificate event occurs https://github.com/jetstack/cert-manager/blob/8dc603e7f5ef64288478b2e7a769a5415ae54ab0/pkg/controller/certificates/issuing/issuing_controller.go#L96
2. A CertificateRequest event occurs who is owed by a Certificate 
https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/issuing/issuing_controller.go#L97-L99
3. A Secret event happened where the Secret name is the same as the Certificate's `status.extPrivateKeySecretName` https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/issuing/issuing_controller.go#L100-L105
4. A Secret event happened who's name is that same as the Certificate's `spec.secretName` https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/issuing/issuing_controller.go#L106-L110

A Secret or Certificate event occurs when those resource objects observe some kind of API call (UPDATE/CREATE/DELETE/PATCH). Based on those predicates, the Certificate(s) will be added to the rate limiting work queue. When that item (Certificate) has been reconciled it will be removed from the queue, unless an error occurred, in which case it will be re-added to the work queue in a rate limited manor (backs off after multiple errors from reconciles).

> If a secret is malformed and we reach this branch: there will be a log statement in the controller output (assuming log level includes error) and this log statement would repeat on each loop?

If data exists in the X.509 certificate Secret Data key, but could not be parse, an error message will be logged like the following:

```
`"failed to determine whether the SecretTemplate matches Secret" error="failed to determine target Secret's base Annotations: error parsing TLS certificate: x509: malformed certificate"`
```

We return nil here so we don't cause another reconcile of the Certificate. The trigger controller should also observe that the certificate is malformed, and cause another issuance. This would then result in the issuance code path of this controller.

Reply:
Excellent explanation, thanks.

> We return nil here so we don't cause another reconcile of the Certificate. The trigger controller should also observe that the certificate is malformed, and cause another issuance. This would then result in the issuance code path of this controller. 

So this would be logged only once, correct?

Reply:
Indeed, we would only expect this to be logged once